File: 2018/day9/marbles.ts
interface IMarble {
	previous: IMarble;
	next: IMarble;
	value: number;
}

interface IPlayer {
	id: number;
	score: number;
}

export class Marbles {
	private players: number;
	private player: number = 0;
	private marbles: number;
	private score: Map<number, number>;
	private marble: IMarble;

	constructor(players: number, lastMarbleWorth: number) {
		this.players = players;
		this.marbles = lastMarbleWorth;
		this.score = new Map();
		this.marble = {
			previous: this.marble,
			next: this.marble,
			value: 0
		};
		this.marble.previous = this.marble;
		this.marble.next = this.marble;
	}

	public play(): void {
		for (let i: number = 1; i <= this.marbles; i++) {
			if (i % 23 === 0) {
				// add to score
				let score: number = this.score.get(this.player) || 0;
				score += i;
				for (const _ of new Array(7)) {
					this.marble = this.marble.previous;
				}
				score += this.marble.value;
				this.score.set(this.player, score);

				this.marble.previous.next = this.marble.next;
				this.marble.next.previous = this.marble.previous;
				this.marble = this.marble.next;
			} else {
				// add to circle
				for (const _ of new Array(1)) {
					this.marble = this.marble.next;
				}
				const m: IMarble = {
					previous: this.marble,
					next: this.marble.next,
					value: i
				};
				this.marble.next.previous = m;
				this.marble.next = m;
				this.marble = m;
			}
			this.player++;
			if (this.player === this.players) {
				this.player = 0;
			}
		}
	}

	public highscore(): IPlayer {
		let highest: IPlayer = {
			id: -1,
			score: -1
		};
		for (const [id, score] of this.score) {
			if (score > highest.score) {
				highest = { id, score };
			}
		}
		return highest;
	}
}

File: 2018/day9-csharp/Marbles.cs
using System.Collections.Generic;

namespace day9
{
	class Marbles
	{
		int players;
		int player = 0;
		int marbles;
		Marble marble;
		Dictionary<int, long> score = new Dictionary<int, long>();

		public Marbles(int players, int marbles)
		{
			this.players = players;
			this.marbles = marbles;
			marble = new Marble()
			{
				Value = 0
			};
			marble.Previous = marble;
			marble.Next = marble;
		}

		public void Play()
		{
			for (int i = 1; i <= this.marbles; i++)
			{
				if (i % 23 == 0)
				{
					// add to score
					long score = this.score.TryGetValue(this.player, out score) ? score : 0;
					score += i;
					for (int j = 0; j < 7; j++)
					{
						this.marble = this.marble.Previous;
					}
					score += this.marble.Value;
					this.score[this.player] = score;

					// Remove marble
					this.marble.Previous.Next = this.marble.Next;
					this.marble.Next.Previous = this.marble.Previous;
					this.marble = this.marble.Next;
				}
				else
				{
					// add to circle
					for (int j = 0; j < 1; j++)
					{
						this.marble = this.marble.Next;
					}
					Marble m = new Marble()
					{
						Previous = this.marble,
							Next = this.marble.Next,
							Value = i
					};
					this.marble.Next.Previous = m;
					this.marble.Next = m;
					this.marble = m;
				}
				this.player++;
				if (this.player == this.players)
				{
					this.player = 0;
				}
			}
		}

		public(int Player, long Score) Highscore()
		{
			(int, long Value) top = (-1, -1);
			foreach (var kv in this.score)
			{
				if (kv.Value > top.Value)
				{
					top = (kv.Key, kv.Value);
				}
			}
			return top;
		}
	}
}
